Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift

```diff
@@ -356,4 +357,100 @@ class HTTPServerPipelineHandlerTest: XCTestCase {
                         .channelRead(HTTPServerRequestPart.end(nil)),
                         .halfClose])
     }
+
+    func testRecursiveChannelReadInvocationsDoNotCauseIssues() throws {
+        func makeRequestHead(uri: String) -> HTTPRequestHead {
+            var requestHead = HTTPRequestHead(version: .init(major: 1, minor: 1), method: .GET, uri: uri)
+            requestHead.headers.add(name: "Host", value: "example.com")
+            return requestHead
+        }
+
+        class VerifyOrderHandler: ChannelInboundHandler {
+            typealias InboundIn = HTTPServerRequestPart
+            typealias OutboundOut = HTTPServerResponsePart
+
+            enum NextExpectedMessageType {
+                case head
+                case end
+            }
+            enum State {
+                case req1HeadExpected
+                case req1EndExpected
+                case req2HeadExpected
+                case req2EndExpected
+                case req3HeadExpected
+                case req3EndExpected
+                case reqBoomHeadExpected
+                case reqBoomEndExpected
+
+                case done
+            }
+            var state: State = .req1HeadExpected
+
+            func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+                let req = self.unwrapInboundIn(data)
+                switch req {
+                case .head(let head):
+                    var sendEnd = true
```

Comment:
This doesn't look right. It may *be* right, but it's probably better to place this in the relevant switch cases.

Reply:
I think it is. We only delay the `.end` for `/req_1`